alf/optimizers/optimizers.py

```diff
@@ -192,6 +201,80 @@ def __init__(self,
             NewCls.counter += 1
 
+    @common.add_method(NewCls)
+    def _clone_params(self, capacity_ratio: float):
+        """clone the parameters
+
+        Args:
+            capacity_ratio: the capacity ration specifying the ratio of learnable parameters
+                to the number of all parameters. Only copy if ``capacity_ratio < 1``
+                and ``masked_out_value`` is None.
+        Returns:
+            an empty dictionary if ``self._masked_out_value`` is None; otherwise, return
+            a dictionary with the parameter as the key and the current parameter value as
+            the value of the dictionary.
+        """
+        param_values = {}
+        if capacity_ratio < 1 and self._masked_out_value is None:
+            for param_group in self.param_groups:
+                for p in param_group['params']:
+                    # only save previous param value if masked_out_value is unspecified
+                    param_values[p] = p.data.clone()
+        return param_values
+
+    @common.add_method(NewCls)
+    def _adjust_capacity(self, capacity_ratio: float, old_param_values: Dict):
+        """adjust the capacity of by apply reassigning old parameter values to the
+        parameter according to the capacity mask.
+
+        Args:
+            capacity_ratio: the capacity ration specifying the ratio of learnable parameters
+                to the number of all parameters
+            old_param_values: a dictionary with the parameter as the key and the parameter
+                value (e.g. from the previous iteration)  as the dictionary value.
+                These values will be used to reset the corresponding parameter values
+                after each gradient step if the corresonding capacity mask is True,
+                effectively excluding it from learning. 
+        """
+        if capacity_ratio < 1:
+            # To achieve this, we assign a random number for each element of
+            # the parameter. An element is turned on if its assigned random number
+            # is less than capacity_ratio. To save memory, we don't store the
+            # random numbers. Instead, we save the random number generator state.
+            org_rng_state = common.get_torch_rng_state()
+
+            for param_group in self.param_groups:
+                for p in param_group['params']:
+                    state = self.state[p]
+                    old_param_val = old_param_values.get(p, None)
+                    # get, save and set random number generator state
+                    if 'rng_state' not in state:
+                        # record random number generator state in ``self.state``
+                        state['rng_state'] = common.get_torch_rng_state(
+                            self._random_number_generator.device)
+                    else:
+                        rng_state = state['rng_state']
+                        common.set_torch_rng_state(*rng_state)
+
```

Comment:
`self._random_number_generator.set_state()`

I wonder how the unittest can pass.

Reply:
> `self._random_number_generator.set_state()`
> 
> I wonder how the unittest can pass.

The reason was that the unittests are designed for the previous version before using rnd as member attributes. 
Updated the unittests and also in optimizers.